### PR TITLE
Return better messages for JSON errors in API.

### PR DESCRIPTION
**Pull request by dennisoelkers**
## Motivation and Context

This PR is aiming to improve error messages returned by the API in case of JSON mapping errors. Previously, error messages caused by deserialization failures were of little help, because a) they were using Java fqcn's and b) they were too noisy. This PR is changing this to return a) the location of the error in the JSON input and b) the path to the property that failed to deserialize in a structured and concise way.

### For unexpected type:

Before:
```
{
	"type": "ApiError",
	"message": "Cannot deserialize value of type `int` from String \"foo\": not a valid `int` value\n at [Source: (org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$UnCloseableInputStream); line: 11, column: 13] (through reference chain: org.graylog.plugins.views.search.rest.AutoValue_SearchDTO$Builder[\"queries\"]->java.util.LinkedHashSet[0]->org.graylog.plugins.views.search.rest.AutoValue_QueryDTO$Builder[\"timerange\"]->org.graylog2.plugin.indexer.searches.timeranges.AutoValue_RelativeRange$Builder[\"from\"])"
}
```

After:
```
{
	"type": "RequestError",
	"message": "Error at \"queries.[0].timerange.from\" [11, 13]: Must be of type int",
	"line": 11,
	"column": 13,
	"path": "queries.[0].timerange.from",
	"reference_path": "org.graylog.plugins.views.search.rest.AutoValue_SearchDTO$Builder[\"queries\"]->java.util.LinkedHashSet[0]->org.graylog.plugins.views.search.rest.AutoValue_QueryDTO$Builder[\"timerange\"]->org.graylog2.plugin.indexer.searches.timeranges.AutoValue_RelativeRange$Builder[\"from\"]"
}
```

### For generic JSON parsing (superfluous `,`):

Before:
```
{
	"type": "ApiError",
	"message": "Unexpected character ('}' (code 125)): was expecting double-quote to start field name\n at [Source: (org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$UnCloseableInputStream); line: 12, column: 5] (through reference chain: org.graylog.plugins.views.search.rest.AutoValue_SearchDTO$Builder[\"queries\"]->java.util.LinkedHashSet[0]->org.graylog.plugins.views.search.rest.AutoValue_QueryDTO$Builder[\"timerange\"])"
}
```

After:
```
{
	"type": "RequestError",
	"message": "Error at \"queries.[0].timerange\" [12, 4]: Unexpected character ('}' (code 125)): was expecting double-quote to start field name\n at [Source: (org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$UnCloseableInputStream); line: 12, column: 4]",
	"line": 12,
	"column": 4,
	"path": "queries.[0].timerange",
	"reference_path": "org.graylog.plugins.views.search.rest.AutoValue_SearchDTO$Builder[\"queries\"]->java.util.LinkedHashSet[0]->org.graylog.plugins.views.search.rest.AutoValue_QueryDTO$Builder[\"timerange\"]"
}
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: changelog/unreleased/pr-18367.toml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+type = "a"
+message = "Improving API error responses for JSON parsing failures.."
```
Comment (bernd) [inline changelog/unreleased/pr-18367.toml:2]: Double full stop.
```suggestion
message = "Improving API error responses for JSON parsing failures."
```

File: graylog2-server/src/main/java/org/graylog2/plugin/rest/RequestError.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import com.google.auto.value.AutoValue;
+
+@JsonAutoDetect
+@AutoValue
+@JsonTypeName("RequestError") // Explicitly indicates the class type to avoid AutoValue_ at the beginning
+public abstract class RequestError implements GenericError {
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/plugin/rest/RequestError.java:27]: Let's note in `UPGRADING.md` the change of the API error response `type` field from `ApiError` to `RequestError` for these errors.

File: graylog2-server/src/main/java/org/graylog2/shared/rest/exceptionmappers/JsonMappingExceptionMapper.java [hunk 3]
```diff
@@ -31,8 +33,44 @@
 public class JsonMappingExceptionMapper implements ExceptionMapper<JsonMappingException> {
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/shared/rest/exceptionmappers/JsonMappingExceptionMapper.java:33]: Let's include a test for the exception mapper.

File: changelog/unreleased/pr-18367.toml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+type = "a"
```
Comment (bernd) [inline changelog/unreleased/pr-18367.toml:1]: Let's make it a `c` because it's a change in the API error responses.

**Review by bernd — COMMENTED** [review]
Thank you! I have some comments.

The new error message no longer includes any type information about which class the issue is about.

Example:
```
Cannot construct instance of `org.graylog.plugins.views.search.searchtypes.events.EventList$Builder`
```

That information doesn't seem to be part of the response anymore. I was using the information as a starting point for debugging.

Can we bring that information back?

File: graylog2-server/src/main/java/org/graylog2/plugin/rest/RequestError.java (hunk 2 above)
Comment (dennisoelkers) [reply graylog2-server/src/main/java/org/graylog2/plugin/rest/RequestError.java:32]: :white_check_mark:

File: changelog/unreleased/pr-18367.toml (hunk 4 above)
Comment (dennisoelkers) [reply changelog/unreleased/pr-18367.toml]: :white_check_mark:

File: changelog/unreleased/pr-18367.toml (hunk 1 above)
Comment (dennisoelkers) [reply changelog/unreleased/pr-18367.toml]: :whi

**Review by bernd — APPROVED** [review]
Thank you!

**Pull request merged**